### PR TITLE
PolyBFT consensus protocol

### PR DESCRIPTION
**Pull request by Stefan-Ethernal**
# Description

Placeholder PR for PolyBFT consensus protocol epic.

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [x] I have added sufficient documentation in code

## Testing

- [ ] I have tested this code with the official test suite
- [ ] I have tested this code manually

### Manual tests

Please complete this section if you ran manual tests for this functionality, otherwise delete it

### CODE REVIEW

**Comment by vcastellm** [comment]
Let's mark this Ready for review

File: contracts/system_addresses.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+	SystemCaller = types.StringToAddress("0xffffFFFfFFffffffffffffffFfFFFfffFFFfFFfE")
+	// L2StateSender is an address of bridge contract to the rootchain
+	L2StateSenderContract = types.StringToAddress("0x1002")
+
+	// NativeTransferPrecompile is an address of native transfer precompile
+	NativeTransferPrecompile = types.StringToAddress("0x2020")
+	// BLSAggSigsVerificationPrecompile is an address of BLS aggregated signatures verificatin precompile
+	BLSAggSigsVerificationPrecompile = types.StringToAddress("0x2030")
+	// ConsolePrecompile is and address of Hardhat console precompile
+	ConsolePrecompile = types.StringToAddress("0x000000000000000000636F6e736F6c652e6c6f67")
```
Comment (vcastellm) [inline contracts/system_addresses.go:26]: Do we want to leave this?

File: Makefile [hunk 2]
```diff
@@ -11,10 +11,11 @@ bindata:
 .PHONY: protoc
 protoc:
 	protoc --go_out=. --go-grpc_out=. ./server/proto/*.proto
-	protoc --go_out=. --go-grpc_out=. ./protocol/proto/*.proto
+	#protoc --go_out=. --go-grpc_out=. ./protocol/proto/*.proto
```
Comment (vcastellm) [inline Makefile:14]: This comment

File: blockchain/blockchain.go [hunk 3]
```diff
@@ -679,44 +679,48 @@ func (b *Blockchain) WriteHeadersWithBodies(headers []*types.Header) error {
...
-	return b.verifyBlock(block)
+	_, err := b.verifyBlock(block)
+
+	return err
 }
 
 // VerifyFinalizedBlock verifies that the block is valid by performing a series of checks.
 // It is assumed that the block status is sealed (committed)
-func (b *Blockchain) VerifyFinalizedBlock(block *types.Block) error {
+func (b *Blockchain) VerifyFinalizedBlock(block *types.Block) (*types.FullBlock, error) {
```
Comment (vcastellm) [inline blockchain/blockchain.go:689]: Something to keep an 👁️

File: state/immutable-trie/snapshot.go [hunk 4]
```diff
@@ -72,7 +74,76 @@ func (s *Snapshot) GetCode(hash types.Hash) ([]byte, bool) {
 }
 
 func (s *Snapshot) Commit(objs []*state.Object) (state.Snapshot, []byte) {
```
Comment (vcastellm) [inline state/immutable-trie/snapshot.go:76]: Watch out this

File: types/rlp_marshal.go [hunk 5]
```diff
@@ -186,5 +190,13 @@ func (t *Transaction) MarshalRLPWith(arena *fastrlp.Arena) *fastrlp.Value {
 	vv.Set(arena.NewBigInt(t.R))
 	vv.Set(arena.NewBigInt(t.S))
 
+	if !t.IsLegacyTx() {
+		vv.Set(arena.NewBytes([]byte{byte(t.Type)}))
+	}
+
+	if t.IsStateTx() {
```
Comment (vcastellm) [inline types/rlp_marshal.go]: In theory this shouldn't affect `ibft` as there are not state txns

**Review by vcastellm — COMMENTED** [review]
Overall I just see some possible interactions with old code to keep an eye but nothing major.

**Review by vcastellm — APPROVED** [review]

File: contracts/system_addresses.go (hunk 1 above)
Comment (Stefan-Ethernal) [reply contracts/system_addresses.go:26]: I think it is convenient to have console precompile enabled (though it is coupled to the development environment strictly). We may need a broader understanding of how we will draw lines between production vs staging vs development environment in Polygon Edge in the future.

**Review by goran-ethernal — APPROVED** [review]

**Review by Nemanja0x — APPROVED** [review]

**Review by igorcrevar — APPROVED** [review]

**Pull request merged**